src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs

```diff
@@ -0,0 +1,195 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.Text;
+
+#nullable enable
+
+namespace System.Reflection
+{
+    internal partial struct TypeNameParser
+    {
+#if !MONO // Mono never needs unescaped names
+        private const char EscapeCharacter = '\\';
+
+        /// <summary>
+        /// Removes escape characters from the string (if there were any found).
+        /// </summary>
+        private static string UnescapeTypeName(string name)
+        {
+            int indexOfEscapeCharacter = name.IndexOf(EscapeCharacter);
+            if (indexOfEscapeCharacter < 0)
+            {
+                return name;
+            }
+
+            // this code path is executed very rarely (IL Emit or pure IL with chars not allowed in C# or F#)
+            var sb = new ValueStringBuilder(stackalloc char[64]);
+            sb.Append(name.AsSpan(0, indexOfEscapeCharacter));
+
+            for (int i = indexOfEscapeCharacter; i < name.Length;)
+            {
+                char c = name[i++];
+
+                if (c != EscapeCharacter)
+                {
+                    sb.Append(c);
+                }
+                else if (i < name.Length && name[i] == EscapeCharacter) // escaped escape character ;)
+                {
+                    sb.Append(c);
+                    // Consume the escaped escape character, it's important for edge cases
+                    // like escaped escape character followed by another escaped char (example: "\\\\\\+")
+                    i++;
+                }
+            }
+
+            return sb.ToString();
```

Comment:
```suggestion
            return Unescape(name, indexOfEscapeCharacter);

            static string Unescape(string name, int indexOfEscapeCharacter)
            {
                // this code path is executed very rarely (IL Emit or pure IL with chars not allowed in C# or F#)
                var sb = new ValueStringBuilder(stackalloc char[64]);
                sb.EnsureCapacity(name.Length);
                sb.Append(name.AsSpan(0, indexOfEscapeCharacter));

                for (int i = indexOfEscapeCharacter; i < name.Length;)
                {
                    char c = name[i++];

                    if (c != EscapeCharacter)
                    {
                        sb.Append(c);
                    }
                    else if (i < name.Length && name[i] == EscapeCharacter) // escaped escape character ;)
                    {
                        sb.Append(c);
                        // Consume the escaped escape character, it's important for edge cases
                        // like escaped escape character followed by another escaped char (example: "\\\\\\+")
                        i++;
                    }
                }

                return sb.ToString();
            }
```
It may be nice to outline the rare path given that it has an expensive stackalloc frame.